src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const statesHistory = [];
+
+  actions.forEach((action) => {
+    let nextState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        nextState = {};
+        break;
+
+      case 'addProperties':
+        nextState = { ...nextState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          delete nextState[key];
+        });
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
```

Comment:
The use of a default case that throws an error for unknown action types is a good practice for catching unexpected inputs. However, ensure that the actions are validated before being passed to this function to prevent runtime errors.